test/externalTests/external_test.py

```diff
@@ -0,0 +1,364 @@
+            evm_version=evm_version,
+            via_ir="true",
+            optimizer="true",
+            yul="true"
+        )
+    }
+    return json.loads(switch.get(preset))
+
+
+def run_cmd(command: str, env: dict = None, logfile: str = None) -> int:
+    """
+    Args:
+        command: command to run
+        logfile: log file name
+        env:     dictionary holding key-value pairs for bash environment variables
+
+    Returns:
+        int: The exit status of the command. Exit status codes are:
+            0     -> Success
+            1-255 -> Failure
+    """
+    if not logfile:
+        logfile = os.devnull
+    if not env:
+        env = os.environ.copy()
+    with open(
+        file=logfile,
+        mode='w',
+        encoding='utf8'
+    ) as log:
+        ret = subprocess.run(
+            command,
+            shell=True,
+            check=True,
+            executable='/bin/bash',
+            env=env,
+            stdout=log if not logfile else None,
+            stderr=None
+        )
+        return ret.returncode
+
+
+@dataclass
+class SolcConfig:
+    binary_type: str
+    binary_path: str
+    branch: Optional[str] = field(default="master")
+    install_dir: Optional[str] = field(default="solc")
+    solcjs_src_dir: Optional[str] = field(default="")
+
+
+@dataclass
+class TestConfig:
+    repo_url: str
+    ref_type: str
+    ref: str
+    config_file: Optional[str]
+    config_var: Optional[str]
+    build_dependency: Optional[str] = field(default="nodejs")
+    compile_only_presets: Optional[List[str]] = field(default_factory=list)
+    settings_presets: Optional[List[str]] = field(default_factory=list)
+    evm_version: Optional[str] = field(default=CURRENT_EVM_VERSION)
+    solc: Dict[str, SolcConfig] = field(
+        default_factory=lambda: defaultdict(SolcConfig))
+
+    def __post_init__(self):
+        if isinstance(self.solc, dict):
+            self.solc = SolcConfig(**self.solc)
+
+    def selected_presets(self):
+        return self.compile_only_presets + self.settings_presets
+
+
+class TestRunner(metaclass=ABCMeta):
+
+    @staticmethod
+    def on_local_test_dir(fn):
+        """Run function inside the test directory"""
+
+        def f(self, *args, **kwargs):
+            if self.test_dir:
+                os.chdir(self.test_dir)
+            else:
+                raise RuntimeError("Test directory not defined")
+            return fn(self, *args, **kwargs)
+        return f
+
+    @abstractmethod
+    def setup_environment(self, test_dir: Path):
+        pass
+
+    @abstractmethod
+    def clean(self):
+        pass
+
+    @abstractmethod
+    def compiler_settings(self, solc_version: str, presets: Tuple[str] = AVAILABLE_PRESETS):
+        pass
+
+    @abstractmethod
+    def compile(self, solc_version: str, preset: str):
+        pass
+
+    @abstractmethod
+    def run_test(self, preset: str):
+        pass
+
+
+class ExternalTest:
```

Comment:
I think this would make more sense a module with free functions. This is not really a test as the name implies. Just a bunch of utilities for setting up the environment for a test. They don't even all share state. `config` would be better off being passed in as a parameter IMO.

Reply:
Sounds good for me as well. I will change it. Any suggestion for the name of the module?

Reply:
Could be just `common.py` if it stays in this directory. Alternatively, something like `exttest_setup.py` or `setup_helpers.py` would be good too.